src/monster.cpp

```diff
@@ -2393,6 +2393,17 @@ void monster::drop_items_on_death()
 
 void monster::process_one_effect( effect &it, bool is_new )
 {
+    // if effect has source try figuring out who to credit
+    const auto &eff_source = it.get_source().get_character_id();
+    Creature *source_creature = nullptr;
+    if( !eff_source ) {
+        source_creature = nullptr;
+    } else if( eff_source.value() == get_player_character().getID() ) {
+        source_creature = &get_player_character();
+    } else {
+        source_creature = g->find_npc( eff_source.value() );
+    }
+
```

Comment:
This logic should probably be encapsulated in a member function of `effect_source`.